source/ada/lsp-lal_utils.adb

```diff
@@ -66,4 +70,132 @@ package body LSP.Lal_Utils is
+      Source_Units : Analysis_Unit_Array (Sources'Range);
+   begin
+      for N in Sources'Range loop
+         Source_Units (N) := Context.Get_From_File
+           (Sources (N).Display_Full_Name,
+            Charset => Charset);
+      end loop;
+
+      declare
+         References : constant Ada_Node_Array :=
+           Definition.P_Find_All_References (Source_Units);
+      begin
+         if Include_Definition then
+            return References & (1 => Definition.As_Ada_Node);
+         else
+            return References;
+         end if;
+      end;
+   end Find_All_References;
+
+   -----------------------
+   -- Containing_Entity --
+   -----------------------
+
+   function Containing_Entity (Ref : Ada_Node) return Defining_Name is
+      Parents : constant Ada_Node_Array := Ref.Parents;
+   begin
+      for Parent of Parents loop
+         if Parent.Kind in Ada_Subp_Decl
+                         | Ada_Subp_Body
+                         | Ada_Task_Def
+                         | Ada_Task_Body
+                         | Ada_Package_Body
+                         | Ada_Package_Decl
+         then
+            return Parent.As_Basic_Decl.P_Canonical_Part.P_Defining_Name;
+         end if;
+      end loop;
+
+      return No_Defining_Name;
+   end Containing_Entity;
+
+   ------------------
+   -- Is_Called_By --
+   ------------------
+
+   function Is_Called_By
+     (Name_Node : Name;
+      Sources   : GNATCOLL.VFS.File_Array_Access;
+      Charset   : String)
+      return References_By_Subprogram.Map
+   is
+      use References_By_Subprogram;
+      use References_List;
+      Result     : Map;
+      Definition : Defining_Name;
+   begin
+      if Name_Node = No_Name then
+         return Result;
+      end if;
+
+      --  Attempt to resolve the name, return no results if we can't or if the
+      --  name does not resolve to a subprogram.
+      Definition := Resolve_Name (Name_Node);
+
+      if Definition = No_Defining_Name
+        or else Definition.P_Basic_Decl.Kind not in Ada_Subp_Decl
+      then
+         return Result;
+      end if;
+
+      --  Obtain all the references
+
+      declare
+         Refs : constant Ada_Node_Array := Find_All_References
+           (Definition, Sources, Charset, Include_Definition => False);
+         Containing : Defining_Name;
+      begin
+         --  Go through all references to Name, organising them by containing
+         --  subprogram.
+
+         for Ref of Refs loop
+            --  Only consider references that are calls.
+            --  ??? To be discussed: how do we want to handle an access to
+            --  a subprogram?
+            if Ref.As_Name.P_Is_Call then
+               --  We have a reference, and this a call: find the containing
+               --  subprogram or task
+               Containing := Containing_Entity (Ref);
+
+               if Containing /= No_Defining_Name then
+                  --  ??? Question: is it possible for the containing
```

Comment:
What is the question ? I don't understand

Reply:
thinko. Will remove.